psycopg/psycopg/connection.py

```diff
@@ -851,6 +865,42 @@ def notifies(self) -> Generator[Notify, None, None]:
                 n = Notify(pgn.relname.decode(enc), pgn.extra.decode(enc), pgn.be_pid)
                 yield n
 
+    @contextmanager
+    def pipeline(self) -> Iterator[Pipeline]:
+        """Context manager to switch the connection into pipeline mode."""
+        with self.lock:
+            if self._pipeline is None:
+                # We must enter pipeline mode: create a new one
+                pipeline = self._pipeline = Pipeline(self.pgconn)
+            else:
+                # we are already in pipeline mode: bail out as soon as we
+                # leave the lock block.
+                pipeline = None
+
+        if not pipeline:
+            # No-op re-entered inner pipeline block.
+            yield self._pipeline
+            return
+
+        try:
+            with pipeline:
+                try:
+                    yield pipeline
+                finally:
+                    with self.lock:
+                        pipeline.sync()
+                        try:
+                            # Send an pending commands (e.g. COMMIT or Sync);
+                            # while processing results, we might get errors...
+                            self.wait(pipeline._communicate_gen())
```

Comment:
Talking about sync/async pipelines: maybe instead it would be better to move this code into the pipelines [a]exit? It uses private-ish pipeline methods and public connection ones...

Reply:
What do you think about https://github.com/psycopg/psycopg/commit/fbb027c0f355ae0f2a02258b8795d9819cdeeae2 ?

Reply:
But then we need to distinguish back `Pipeline` and `AsyncPipeline` thus making 5465b9248d331fc9871adc7a61ce88db08417aa1 pointless?

Reply:
Yes, I've based this commit on the commit before dropping sync/async pipeline. This change starts making the distinction of the two classes meaningful.

Reply:
Ok; sounds good, pushed.